### PR TITLE
Remove/red bubble data from window state

### PR DESCRIPTION
**Pull request by CodeyGuyDylan**
## Proposed changes:

There are several small changes made here to deal with the async loading of red bubble notifications on the front end, I will do my best to summarize them below

* Remove red_bubble_notifications from the initial window state. This call was bypassing cache and calling several HTTP requests on every page load to ensure we had the most up-to-date info. By removing them from the initial state and calling them async on the frontend, this should improve initial pageload speed
* Several loading states had to be updated to ensure no errors or weird states while the red bubble data loaded. You'll see several changes in the code to how the data is handled before the red bubble data has been fetched
* Added check for cookie on the frontend as well as the backend as the REST API call for the red bubble data did not send all cookies along with it, and therefore was returning incorrect data when a notice had been dismissed. Rather than increase security risks, I figured checking for the cookie on the frontend as well would be the easiest and safest way to do this
* Moved all red bubble related functions and constants to a new `Red_Bubble_Notifications` class to further clean up the initializer class and better segment the data on the backend

NOTE: The reason this change is being made, even though we are still calling the red bubble function on the backend, is because when we call the function on the backend, we are using a transient cache that is set for 1 hour. The only non-cached response we were calling was the value sent to the initial window state. By removing that instance, the red bubble data should only block page load once per hour rather than every page as it is currently doing

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

P2: pghfsA-df-p2

## Does this pull request change what data or activity we track or use?

No

### Main things to test

There is a lot to test here, and I will try to cover everything important. However, feel free to run additional tests after inspecting the code if you think there could be situations I've missed testing to ensure we don't break anything with this. This is a very large PR, and semi-high risk due to the nature of changing how data is fetched. So the more thoroughly it can be tested, the better

1. Checkout this branch via the Jetpack Beta plugin or your local dev environment, make sure you install the Jetpack Debug Helper if you are using a JN site

NOTE: If you want to test the bad installation notice specifically, you will have to use your local environment to test, as I don't know how to trigger that error without editing the code in the plugin. I will include how to test that last in the instructions

2. Before connecting your site or user account, install Query Monitor to ensure we aren't calling any more HTTP requests on the backend than necessary
3. Go to My Jetpack and open the devtools to ensure the red-bubble-notifications are being loaded correctly in the REST API call. Also ensure you see the welcome banner
image
image

Optional: Add a new user with the editor permissions and test from another browser to ensure none of these steps are broken for non-admins either

4. Connect your site
5. Purchase a Complete plan on your site however you'd like
6. During the backup, disconnect your site to ensure the backup fails
7. Go to the Store Admin and change the expiration date on your Complete plan to be a few days in the future
8. In the Jetpack Debug Helper, Enable the Jetpack Scan helper
9. In the Jetpack Scan helper, add 5 EICAR threats
image
10. Run a scan on the site to detect the threats

This should give you all the notices that are possible outside of the Bad Install notice. I will get to that later

NOTE: I had a very hard time getting a backup to fail and for the error to be returned correctly on a Jurassic Ninja site 🤔 I am not sure why, but I don't think this has to do with the changes in this PR because I am able to get it to show on the local docker environment pretty easy. I think maybe the backup data has some cache that it's not returning the failed backup quickly, I am not entirely sure, but feel free to test this on your local environment to see that it works

11. For each notice, refresh the page to ensure everything loads without error. Then dismiss the notice, and repeat with the next one to ensure all notices can load with the async red bubble data
image
image
image
image

12. Disconnect your user account and site connection. Ensure you see the site connection banner, refresh to make sure it loads okay
image
13. Connect your site but not user account. Ensure the user connection banner loads too
image

#### Welcome Banner

1. Fill out the welcome banner and ensure the evaluation recommendations show up as expected
image
2. Click the ellipses and select Redo and make sure the banner lets you redo the survey
image
3. Dismiss the welcome banner however you'd like and make sure it stays gone on refresh
image

#### Bad install notice

1. On your local environment, make sure you have Jetpack Protect installed and active
5. Go to `projects/plugins/protect/jetpack-protect.php`
6. Update line 47 to
```php
if ( ! is_readable( $jetpack_autoloader ) ) {
```
7. Go to My Jetpack and ensure the Bad installation notice displays correctly
image

#### Tracks events

1. On initial pageload of My Jetpack, ensure that the `jetpack_myjetpack_page_view` event is only loaded once, and that it includes all red bubble slugs
 
Screenshot 2025-03-07 at 1 50 47 PM

2. On initial pageload of My Jetpack, ensure that each product card only fires the `jetpack_myjetpack_product_card_load` event one time each, for a total of 10

### CODE REVIEW

File: projects/packages/my-jetpack/_inc/components/product-cards-section/videopress-card/index.tsx [hunk 1]
```diff
@@ -33,10 +33,12 @@ const VideopressCard: ProductCardComponent = props => {
 		videoCount,
 	} );
 
-	const customLoadTracks = {
-		stats_period: featuredStats?.period,
-		video_count: videoCount,
-	};
+	const customLoadTracks = useMemo( () => {
```
Comment (CodeyGuyDylan) [inline projects/packages/my-jetpack/_inc/components/product-cards-section/videopress-card/index.tsx:36]: Without being memoized, this was causing the videopress card to be unnecessarily re-rendered, which also caused multiple tracks to fire for its loading event

File: projects/packages/my-jetpack/_inc/components/product-card/index.tsx [hunk 2]
```diff
@@ -132,12 +136,26 @@ const ProductCard: FC< ProductCardProps > = props => {
 	 * Sends an event when the card loads
 	 */
 	useEffect( () => {
+		if ( isTracksFired || isDataLoading || isAllProductsLoading ) {
```
Comment (CodeyGuyDylan) [inline projects/packages/my-jetpack/_inc/components/product-card/index.tsx]: The async data loads were causing this load event to fire multiple times for every card. Most of the extra events had a status of `undefined`. This ensures they are only called once

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-paid-plan-needs-plugin-install-activation-notice.tsx [hunk 3]
```diff
@@ -282,11 +292,13 @@ const usePaidPlanNeedsPluginInstallActivationNotice = ( redBubbleAlerts: RedBubb
 			priority: NOTICE_PRIORITY_MEDIUM + ( isInstallingOrActivating ? 1 : 0 ),
 		};
 
-		setNotice( {
-			title: noticeTitle,
-			message: noticeContent,
-			options: noticeOptions,
-		} );
+		if ( ! isLoading && ! checkForCookie( `${ planSlug }--plugins_needing_installed_dismissed` ) ) {
```
Comment (CodeyGuyDylan) [inline projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-paid-plan-needs-plugin-install-activation-notice.tsx]: We are checking here for cookie as well, since the REST API call doesn't send over the cookies when retrieving the red bubble slugs from the backend. Instead of increasing security risk by sending all cookies along with the request, I thought we could just check for it here as well since it's not much added code

File: projects/packages/my-jetpack/_inc/components/my-jetpack-screen/index.jsx [hunk 4]
```diff
@@ -134,10 +141,24 @@ export default function MyJetpackScreen() {
...
+
+		if (
+			! isRedBubbleAlertsError &&
+			typeof redBubbleAlerts === 'object' &&
+			Object.keys( redBubbleAlerts )?.length
+		) {
+			customTracksData = {
+				red_bubble_alerts: Object.keys( redBubbleAlerts ).join( ',' ),
+			};
+		}
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/components/my-jetpack-screen/index.jsx]: Considering we have a proper type-safe definition of redBubbleAlerts (which we probably should have), we don't need to functionally guard the prop here.

File: projects/packages/my-jetpack/_inc/components/product-card/index.tsx (hunk 2 above)
Comment (robertsreberski) [reply projects/packages/my-jetpack/_inc/components/product-card/index.tsx]: I would just propse to make it clearer why we do it. What would you say about implementing more descriptive variables?

```
  const isDataReady = !isDataLoading && !isAllProductsLoading;
  const shouldTrackEvent = isDataReady && !isTracksFired;
```

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx [hunk 5]
```diff
@@ -171,5 +175,8 @@ export function useGetReadableFailedBackupReason(): ReasonContent {
 		}
 	}, [ status ] );
 
-	return reasonContent;
+	return {
+		isLoading: isRedBubbleAlertsLoading,
+		reasonContent: reasonContent,
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx:173]: From this code I understand the reasonContent always has a structure { title, text } - these two props are not "optional". We can update the type and references to reflect that.

I can see it can be ReactElement | string | null

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx [hunk 6]
```diff
@@ -13,8 +16,9 @@ export type ReasonContent = {
  * @return {ReasonContent} An object containing each tooltip's title and text content.
  */
 export function useGetReadableFailedBackupReason(): ReasonContent {
-	const { backup_failure: backupFailure } =
-		getMyJetpackWindowInitialState( 'redBubbleAlerts' ) || {};
+	const { data: redBubbleAlerts, isLoading: isRedBubbleAlertsLoading } = useRedBubbleQuery();
+
+	const { backup_failure: backupFailure } = redBubbleAlerts || {};
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx]: It would be of course better to have a standardized prop names in snakeCase.

Secondly, I think we could simplify the code here not to depend on "||" fallback values. It doesn't really feel like a TypeScript approach.

File: projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx [hunk 7]
```diff
@@ -88,7 +91,8 @@ const BackupCard = props => {
 
 	return (
 		<ProductCard slug={ productSlug } Description={ isError && noDescription } { ...props }>
-			{ isError && (
+			{ isBackupFailedReasonLoading && <LoadingBlock height="75px" width="100%" /> }
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx:94]: It's great to see these Loading Blocks. One question / suggestion - where does this hardcoded height come from? Could we put it into re-usable constant?

File: projects/packages/my-jetpack/_inc/components/product-cards-section/videopress-card/index.tsx (hunk 1 above)
Comment (robertsreberski) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/videopress-card/index.tsx:36]: That's a good catch!

File: projects/packages/my-jetpack/_inc/data/products/use-products.ts [hunk 8]
```diff
@@ -29,6 +29,10 @@ const refetchProducts = async (
 	const { data: refetchedProducts } = await refetch();
 	const prevProducts = window.myJetpackInitialState.products.items;
 
+	if ( ! refetchedProducts ) {
+		return;
+	}
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/data/products/use-products.ts]: Type definition doesn't suggest this variable can be nullable - we might update that?

File: projects/packages/my-jetpack/_inc/data/welcome-banner/use-welcome-banner.ts [hunk 9]
```diff
@@ -1,20 +1,48 @@
...
 		'isWelcomeBannerVisible',
-		Object.keys( redBubbleAlerts ).includes( 'welcome-banner-active' )
+		redBubbleAlertKeys.includes( 'welcome-banner-active' )
 	);
 
+	useEffect( () => {
+		if (
+			! isRedBubbleAlertsLoading &&
+			! isRedBubbleAlertsError &&
+			redBubbleAlertKeys.includes( 'welcome-banner-active' )
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/data/welcome-banner/use-welcome-banner.ts]: Combined with useMemo() above, it's kind of implicit that isRedBubbleAlertsError || isRedBubbleAlertsLoading === is empty === for sure it doesn't include 'welcome-banner-active'

Maybe one of these hooks is redundant?

On another note, maybe we could enhance useSimpleQuery (or use useQuery) to include `onSuccess` callback, that would allow us to update the state value without any useEffect or useMemo?

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/index.ts [hunk 10]
```diff
@@ -1,24 +1,22 @@
...
+	usePaidPlanNeedsPluginInstallActivationNotice( redBubbleAlerts, isLoading );
+	useProtectThreatsDetectedNotice( redBubbleAlerts, isLoading );
+	useExpiringPlansNotice( redBubbleAlerts, isLoading );
+	useBackupNeedsAttentionNotice( redBubbleAlerts, isLoading );
 	useConnectionErrorsNotice();
-	useSiteConnectionNotice( redBubbleAlerts );
-	useBadInstallNotice( redBubbleAlerts );
+	useSiteConnectionNotice( redBubbleAlerts, isLoading );
+	useBadInstallNotice( redBubbleAlerts, isLoading );
 };
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/index.ts:20]: It feels like we are again passing props to many places. Do you think it would be useful to introduce Context on red bubble query?

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/types.ts [hunk 11]
```diff
@@ -0,0 +1 @@
+export type NoticeHookType = ( redBubbleAlerts: RedBubbleAlerts, isLoading: boolean ) => void;
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/types.ts:1]: If you don't import anything in .ts file, I don't think you need to export. Just a small note.

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-protect-threats-detected-notice.tsx [hunk 12]
```diff
@@ -43,9 +42,9 @@ const useProtectThreatsDetectedNotice = ( redBubbleAlerts: RedBubbleAlerts ) =>
 
 	const onCloseClick = useCallback( () => {
 		createCookie( 'protect_threats_detected_dismissed', 7 );
-		delete redBubbleAlerts.protect_has_threats;
+		delete redBubbleAlerts?.protect_has_threats;
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-protect-threats-detected-notice.tsx:45]: Is it necessary and works well now after migration to react-query.

For sure, it is a bad pattern we could solve, now or in the near future.

File: projects/packages/my-jetpack/changelog/remove-red-bubble-data-from-window-state [hunk 13]
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: changed
+
+Take red bubble notifications out of initial state
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/changelog/remove-red-bubble-data-from-window-state]: Now that I prepare changelogs for releases - our best practice is to write a module name upfront, and end the changelog always with the coma. For example:
`My Jetpack: Move Red Bubble notifications out of myJetpackInitialState.`

It's a small suggestion, though, not related to the PR code itself!

File: projects/packages/my-jetpack/global.d.ts [hunk 14]
```diff
@@ -112,6 +112,47 @@ type BackupStatus =
...
+	[ key: `${ string }-bad-installation` ]: {
+		data: {
+			plugin: string;
+		};
+	};
+	backup_failure?: {
+		type: 'warning' | 'error';
+		data: BackupNeedsAttentionData;
+	};
+	[ key: `${ string }--plan_expired` ]: {
```
Comment (robertsreberski) [inline projects/packages/my-jetpack/global.d.ts]: Now that we touch Red Bubble, let's also take care of improving the typing. E.g. I would make it more precise what values we expect here, rather than using a generic `string`. It could for example be:

```
type ProductSlug = 'backup' | 'scan' | 'videopress' | 'search' | 'anti-spam';

type PlanExpiredAlerts = Record<`${ProductSlug}--plan_expired`, PlanExpirationAlert>;
```

Secondly, yeah, we can use `Record` type here, which could make the typing clearer and integrate better with TS suggestions and best practices:
```
type RedBubbleAlerts = { } & PlanExpiredAlerts;
```

Thirdly, we can define subtypes explicitly rather than nesting object definition within the code. E.g.

```
'missing-connection'?: ConnectionAlert;
```

This all can provide better dev ex when working on the code in the future.

**Review by robertsreberski — CHANGES_REQUESTED** [review]
I love how it all looks so much better and in spirit of type safe codebase. Thank you for working on this!

I have provided feedback to make it even more awesome, I will test it later today functionally as well.

File: projects/packages/my-jetpack/_inc/components/product-card/index.tsx (hunk 2 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/components/product-card/index.tsx]: That's a good improvement, thanks!

File: projects/packages/my-jetpack/_inc/components/my-jetpack-screen/index.jsx (hunk 4 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/components/my-jetpack-screen/index.jsx]: Yes good catch. I initially had this because this was failing if the red bubble query failed, but I've since added a fallback of an empty object so I think we are good to remove it after I tested it 😄

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx (hunk 5 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx:173]: Nice catch!

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx (hunk 6 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx]: Similarly to before, I agree! These were here before I added the {} fallback on the red bubble query. Thank you 😄

Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx]: > It would be of course better to have a standardized prop names in snakeCase.

I agree, but I think i'd wand to change those all at once so I think it may be better to handle this sort of thing when we get to our project to remove the window state object altogether. Does that sound okay?

File: projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx (hunk 7 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx:94]: The hard-coded height was the way I came up with to give flexibility to try and have the block be roughly the same height as the content that it's placeholding. There may be a better way to do this 🤔 Can you think of one? I may want to handle that in a separate PR if we are going to change it

File: projects/packages/my-jetpack/_inc/data/products/use-products.ts (hunk 8 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/data/products/use-products.ts]: That's a good point, I'll update this to check if the refetch is an error or is loading instead

File: projects/packages/my-jetpack/_inc/data/welcome-banner/use-welcome-banner.ts (hunk 9 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/data/welcome-banner/use-welcome-banner.ts:35]: The 
```ts
if ( isRedBubbleAlertsError || isRedBubbleAlertsLoading ) {
	return [];
}
```
check was needed in the `useMemo` hook because the object being updated was not enough to trigger a re-render as that is only triggered if a new object is created. Since we are only changing the same object, the variable was not being updated when it needed to be. Similarly with the useEffect, we need the loading statuses in there in order for it to re-run at the correct time. A workaround for this would be to create a new object every time the redBubble slugs change, but that seemed more hacky than this. 

> On another note, maybe we could enhance useSimpleQuery (or use useQuery) to include onSuccess callback, that would allow us to update the state value without any useEffect or useMemo?

But perhaps your suggestion could fix this 🤔

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/index.ts (hunk 10 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/index.ts:20]: In this specific case, I don't think that it's bad to be passing one more prop, especially since all the hooks will have the same props. I think we could make an argument to create contexts for several points of data. Product data could be in context, purchases, etc. And I think that may be the direction we head in the future. However I'd argue against doing it here to avoid making this PR larger than it already is.

I also like to avoid passing props so I understand your concern here, but I think in this specific case it seems reasonable 🤔

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/types.ts (hunk 11 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/types.ts:1]: I've never heard of this, and it's very interesting 🤔 

However I'm getting a linting error when I try to remove the export

`'NoticeHookType' is defined but never used.eslint[@typescript-eslint/no-unused-vars](https://typescript-eslint.io/rules/no-unused-vars)`

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-protect-threats-detected-notice.tsx (hunk 12 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-protect-threats-detected-notice.tsx:45]: Yes I agree, there are a lot of ways we could improve how the red bubble slugs on the frontend. This project being about moving data from backend blocking http requests to async loading on the frontend, I'm not sure it's the best place to address all these improvements 🤔 Especially seeing as how large this PR is already. I could handle it in a followup task, but I'm still unsure of how well it fits in the scope of this project. On top of the PR being large, this project as a whole is pretty large and I want to be cautious of how many things it covers to prevent it from going on too long. Does that sound reasonable?

File: projects/packages/my-jetpack/_inc/data/welcome-banner/use-welcome-banner.ts (hunk 9 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/_inc/data/welcome-banner/use-welcome-banner.ts]: I've updated this a bit to refetch the alert bubbles after dismissal. Let me know what you think

File: projects/packages/my-jetpack/global.d.ts (hunk 14 above)
Comment (CodeyGuyDylan) [reply projects/packages/my-jetpack/global.d.ts]: I've added typing to this to the best of my ability,  let me know what you think!

File: projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx (hunk 7 above)
Comment (grzegorz-cp) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx:94]: Thanks for creating this block, it's very handy. I used it in the interstitial modal.
In another PR we might add a default value for the height and make the width 100%, but I think those values will depend on where the component is used so the defaults might not add much.

Comment (robertsreberski) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/backup-card/index.jsx:94]: Hmm, right, I just think we could have it in some variable, to give it more context? It's not the blocker for this PR, and great job creating this block, first of all!

File: projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx (hunk 6 above)
Comment (robertsreberski) [reply projects/packages/my-jetpack/_inc/hooks/use-notification-watcher/use-get-readable-failed-backup-reason.tsx]: >  I think it may be better to handle this sort of thing when we get to our project to remove the window state object altogether.

Can you note it down in the project (Linear)?

**Comment by robertsreberski** [comment]
Thanks for the hard work on this, @CodeyGuyDylan! It all tests good, and code looks great after the changes.

**Review by robertsreberski — DISMISSED** [review]
